pkg/networkserver/grpc_deviceregistry.go

```diff
@@ -275,392 +269,1752 @@ func (ns *NetworkServer) Get(ctx context.Context, req *ttnpb.GetEndDeviceRequest
+	if st.HasSetField(left) {
+		if st.Device.FieldIsZero(left) {
+			return nil
+		}
+		return st.ValidateFieldsAreNotZero(right...)
+	}
+	for _, r := range right {
+		if !st.HasSetField(r) || !st.Device.FieldIsZero(r) {
+			continue
+		}
+		if err := st.ValidateFieldIsZero(left); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// ValidateIfZeroThenFunc ensures FieldIsZero(left) -> f(map r -> *ttnpb.EndDevice), for each r in right.
+func (st *setDeviceState) ValidateIfZeroThenFunc(isValid func(map[string]*ttnpb.EndDevice) (bool, string), left string, right ...string) error {
+	if st.HasSetField(left) {
+		if !st.Device.FieldIsZero(left) {
+			return nil
+		}
+		return st.ValidateFields(isValid, right...)
+	}
+	if !st.HasSetField(right...) {
+		return nil
+	}
+	return st.ValidateFields(func(m map[string]*ttnpb.EndDevice) (bool, string) {
+		if !m[left].FieldIsZero(left) {
+			return true, ""
+		}
+		return isValid(m)
+	}, append([]string{left}, right...)...)
+}
+
+// ValidateIfNotZeroThenFunc ensures !FieldIsZero(left) -> f(map r -> *ttnpb.EndDevice), for each r in right.
+func (st *setDeviceState) ValidateIfNotZeroThenFunc(isValid func(map[string]*ttnpb.EndDevice) (bool, string), left string, right ...string) error {
+	if st.HasSetField(left) {
+		if st.Device.FieldIsZero(left) {
+			return nil
+		}
+		return st.ValidateFields(isValid, right...)
+	}
+	if !st.HasSetField(right...) {
+		return nil
+	}
+	return st.ValidateFields(func(m map[string]*ttnpb.EndDevice) (bool, string) {
+		if m[left].FieldIsZero(left) {
+			return true, ""
+		}
+		return isValid(m)
+	}, append([]string{left}, right...)...)
+}
+
+// SetFunc is the function meant to be passed to SetByID.
+func (st *setDeviceState) SetFunc(f func(context.Context, *ttnpb.EndDevice) error) func(context.Context, *ttnpb.EndDevice) (*ttnpb.EndDevice, []string, error) {
+	return func(ctx context.Context, stored *ttnpb.EndDevice) (*ttnpb.EndDevice, []string, error) {
+		for p, shouldBeZero := range st.zeroPaths {
+			if stored.FieldIsZero(p) != shouldBeZero {
+				return nil, nil, newInvalidFieldValueError(p)
+			}
+		}
+		for _, f := range st.onGet {
+			if err := f(stored); err != nil {
+				return nil, nil, err
+			}
+		}
+		if err := f(ctx, stored); err != nil {
+			return nil, nil, err
+		}
+		return st.Device, st.SetFields(), nil
+	}
+}
+
+func newSetDeviceState(dev *ttnpb.EndDevice, paths ...string) *setDeviceState {
+	return &setDeviceState{
+		Device: dev,
+		paths:  paths,
+
+		pathsCache:     make(map[string]bool),
+		extraSetsCache: make(map[string]bool),
+		extraGetsCache: make(map[string]bool),
+
+		zeroPaths: make(map[string]bool),
+	}
+}
+
+func setKeyIsZero(m map[string]*ttnpb.EndDevice, get func(*ttnpb.EndDevice) *ttnpb.KeyEnvelope, path string) bool {
+	if dev, ok := m[path+".key"]; ok {
+		if ke := get(dev); !ke.Key.IsZero() {
+			return false
+		}
+	}
+	if dev, ok := m[path+".encrypted_key"]; ok {
+		if ke := get(dev); len(ke.EncryptedKey) == 0 {
```

Comment:
I think this should be the opposite?

```suggestion
		if ke := get(dev); len(ke.EncryptedKey) != 0 {
```